### ISSUE TITLE
Improved Rename and Show Usage

### ISSUE THREAD

**Issue by eidheim** [cppit/jucipp#251]
At the moment, files that are to be included in rename and show usage needs to be opened and parsed by libclang. Sadly, libclang is too slow to do a full parse of an entire medium or larger sized project. However, there are other options, like using cscope, that might solve this adequately. Feel free to post your opinions here, or suggest other solutions.

**Comment by milleniumbug** [comment]
Many IDEs I know (CLion, MSVS) schedule parsing in the background, with on-disk cache so the reparsing doesn't occur when reopening file or reopening program.

**Comment by eidheim** [comment]
I do not think we can use libclang for this as it can take hours to parse a larger project. Additionally, every time one changes the compile options, a full reparse would be needed. CLion and MSVS solves this by using faster but far less complete parsers than libclang.

**Comment by eidheim** [comment]
The new flag CXTranslationUnit_KeepGoing  (http://clang.llvm.org/doxygen/group__CINDEX__TRANSLATION__UNIT.html#ggab1e4965c1ebe8e41d71e90203a723fe9afda3272e16e2aa499fc265615991f968) might be of use, for instance, parsing sources without including system includes and maybe still get some cursor information. I'll investigate later.

**Comment by Smjert** [comment]
Well i just naively tried to create a translation unit for every file that had corresponding information inside the compilation database of the Krita project (around 5k .cpp and .h, 2327 files effectively parsed).
Just a simple single threaded loop and it took 1 minute and 30 seconds.
Now i would be surprised if traversing the AST to find types, functions and function calls would take much more.
Probably the highest time would be needed to create the database with crossreferences or the like.

**Comment by eidheim** [comment]
@Smjert This is much faster than expected. Are you sure the headers were found when parsing the source files? I seldom see parse times below one second on sources of more complex projects. However, if the header files are not found during parsing, it is much faster.

**Comment by Smjert** [comment]
Nevermind ^^', it was a bit late in the day and i supposed that given that i was querying the TU without crashes (like, trying to have some informations from the cursors) it wasn't null.
Now i rechecked and indeed the translation unit returned is null (so error happened).
I'm sorry i did came to conclusions really too fast.

**Comment by Smjert** [comment]
@eidheim: So i fixed my code and it takes like 2-3 seconds to parse each file, so yes it's really not doable like that.

**Disclaimer: Long Text Ahead!**

While having read a bit more about libclang doc, having looked around on what others do (QtCreator, YouCompleteMe, RTags) i think that there aren't many choices. Kind of 3 choices **in general**:

1) Use/write some text parser/tokenizer which quickly creates cross references. Obviously this will be imprecise due to the "impossibility" of dealing with defines/ifdefs (unless one writes a small preprocessor), and having issues with templates.

2) Manually create an AST, trying to speed it up by skipping non interesting things (which would basically mean doing point 1 plus dealing with an AST, preprocessor and stuff).

3) Use libclang with a "clever" use of precompiled headers.

The third choice is the way to go imho though the question is exactly how:

1) QtCreator compiles TU (.h and .cpp) with the preamble stored on disk, automatically created by the option CXTranslationUnit_PrecompiledPreamble, as soon as you open a file.
But it doesn't seems to reuse .pch coming from already parsed headers or to use Clang at all to Find All References, where it resorts to its own AST.
QtCreator so uses clang for syntax highlighting, go to declaration and diagnostic.
The issue here is that if it had to go with a clang only solution, it would do too much work and not reusing the pch it already precompiled.

2) RTags seems to do something very similar but given that it only uses clang it does it a bit more cleverly (though i'm not 100% sure of the details... the code is a bit hard to follow):
Basically for each source file found in the compilation database it walks all the include paths args on the command line and "i think" it compiles all the headers it finds there. It's not then clear to me if it's RTags itself that knows which pch a source file has to use and reuses them, also checking if the args that compiled that pch are the same or if it's up to the user to pass the "include-pch" command line.

That being said both, done like that, still have a big issue imho, that is space consumption of the pch headers (opening the source files of a project like RTags with QtCreator creates preambles of like 10 to 20MB each, and it can go worse https://github.com/Valloric/YouCompleteMe/issues/1419).

What could be done is to create smaller parts/pch which could be reused, depending also on the compile arguments needed (meaning that some of the pch must be duplicated if there are two sources that includes that same include but have different flags like defines or C/C++ version).
The question is where to stop walking the chain of includes. 
My first idea is probably a bit too crazy: create the full dependency tree of includes and then compile the bottomest, then while going up the tree, chain all the pch. This way each branch is an include chain starting from a source file and the whole tree represents all the needed includes that are compiled with the same flags.
The second step would be to store these pch along with flags in some kind of hashmap i guess, where each key is the include path and the values is a list of pch, with each entry being a pch + flags.
Obviously the best situation would be where all the project is compiled with the same flags, so one would have exactly one pch in that list.

Now the dependency tree can be created by clang itself: if one compiles the TU with some special flags that skips function bodies (and maybe something else?), it could already have the include dependency tree on its hands.

Though this might be crazy because the dependency tree could be pretty big... especially if the project itself has a lot of includes and a complicated tree (not including system includes and library), so not only it could require a lot of memory to keep the entire hashmap in memory (ok it could be offloaded to disk..), but also when a change happen it could potentially trigger a very big reparse.

So maybe the less crazy choice is to do what people would already do by hand: precompile system/library headers only, though not doing one mega pch, but as i described earlier, doing once "for each" header and then reuse them when walking up the tree.

Then watch for any flag/source code change and reparse accordingly (for instance if the user adds a #define or #ifdef before an include, then reparse that specific branch of includes).
This could be maybe also sped up by a simple grepping of the header files to find which ones in the subtree uses that define/ifdef, to avoid unnecessary reparsing.

But without actually implementing it/measuring it first i really don't know what really are the time/space issues..

As a last thing: your solution of using KeepGoing, from what i can understand, would maybe speed it up a bit, but first of all it would only permit you to search for types defined by the project itself, no library/system library type or function, because you would need the full name (namespace, class and arguments); second it could be still quite slow if the project has a lot of its own headers.

**Comment by eidheim** [comment]
Thank you for the write up, and for the interesting research you have done on this issue!

I tried CXTranslationUnit_KeepGoing, but it solves nothing regarding this issue, but the flag will be added next week or so for those who have libclang version >=3.9. However, I did some experimentation on parsing with some headers not included. I found that as long as you have included the type definition of an variable, and type definition and return type of a method, one could probably receive enough information to find these variables and methods. One thought then, is to create a PCH of the include files that have the definitions of the type (and return type) of the variable (or method/function) in question. Then, when parsing the rest of the sources in a project, remove includes that do not have the definitions of the type (and return type for methods). This would be fast enough for performing each time show usage or renaming is performed, though we would need a preference option for especially large projects. The problem with caching PCHs for a longer period is troublesome in quite a few occasions I think. Especially problematic is that one now quite often is switching branches, checkouts commits, or performs rebasing in git repositories.

**Comment by eidheim** [comment]
After some additional thought, one would of course need the parameter type declarations too (their includes). 

Regarding PCH's, they are only valid as long as you do not add or remove a header file, so that is why one would have to remove includes from the source you do not include in the temporary PCH. One must possibly just replace all includes in a source with the includes you have in the temporary PCH.

**Comment by Smjert** [comment]
In the meanwhile i did some experimentation too and unfortunately just parsing the TUs with the bare minimum for walking the includes takes like 500ms per file.
For the whole Krita project (~3000 .cpp) it took 9 minutes. Other solutions, like QtCreator, takes like 30 seconds when starting up the project, then whatever search is almost instant.

Well the idea of using PCHs and creating them is that one has to go the deeper it can, without being unmanageable, so that it's unlikely they change and they can be composed.
So continuing with this idea i would try to parse all the includes present in a source file, opening them and doing the same until you hit something that is not part of the project (it goes out of the project root), then you start to transform these in PCHs (so namely all the external libraries and system libraries).

It's true that when switching things may change, but if you have "small enough" PCHs, you can save a lot of time because it's unlikely that many will change, at least for the library and system ones.


What you propose makes sense though, but then you would not be able to research types that are present in other locations of the source (types that appears in function body? class fields?)
I think it has to be an all or nothing.

I'll do some more experimentation...

**Comment by eidheim** [comment]
The idea I proposed was only to solve rename and show usages across multiple files.

One could also exclude source files in the parsing that in no way would include the declaration of the variable or function that you want to rename or show usages of.

**Comment by eidheim** [comment]
Nevertheless, I think this problem needs some additional thought, as the solution is not self evident at all. But it's great that we consider and try out different options!

**Comment by Smjert** [comment]
> The idea I proposed was only to solve rename and show usages across multiple files.
> 
> One could also exclude source files in the parsing that in no way would include the declaration of the variable or function that you want to rename or show usages of.

Well but you would have to retrieve the full name of the thing you have under the cursor to be sure to be searching the right thing, are you sure it would be enough to use the visible/keyboard typed type you see?

> Nevertheless, I think this problem needs some additional thought, as the solution is not self evident at all. But it's great that we consider and try out different options!

Indeed!

And unfortunately the more i go on the more i think what i propose cannot fully done.. it needs big cuts and approximations.

**Comment by eidheim** [comment]
> Well but you would have to retrieve the full name of the thing you have under the cursor to be sure to be searching the right thing, are you sure it would be enough to use the visible/keyboard typed type you see?

The file from which you initiate the rename/show usage would have to be fully parsed so that you would know which declarations to look for. The parsing tricks are to find usages in unopened files that are not parsed.

**Comment by Smjert** [comment]
> The file from which you initiate the rename/show usage would have to be fully parsed so that you would know which declarations to look for. The parsing tricks are to find usages in unopened files that are not parsed.

Ah yes, i thought one thing and written another. The full name is needed in the unopened files.

I start getting much better what you're thinking and it can definitely work well.
I forgot that libclang can tell you where a type/function is defined.. in this way you could on the fly just do a quick grep of which files contains what you're searching and then parse those just including the single header + KeepGoing; or just parsing all the source files, whichever is the fastest.

**Comment by eidheim** [comment]
I think you understand me even though my explanation was a bit bad, I'll add some examples Tomorrow morning just to be on the safe side. But I'm pretty sure it will not work in all cases, but then, no one has a rename refactoring tool in C++ that work in 100% of the cases, and I'm not sure if it is at all possible. I will be happy with a 95% working solution though!

Actually, you will not need the KeepGoing flag in this case, the flag helps if your source have an include that it does not find for instance. With or without the flag, libclang will create the cursors it can for the entire source that is being parsed. This is good news though, since the solution we are discussing will work on older libclang versions as well.

**Comment by Smjert** [comment]
Well the idea was that the unopened files would artificially have all the includes not to be found except for the one that contains the type you're trying to find, otherwise it would still a no go if you had to entirely parse the source files where you think the type is referenced.
For instance i've tried again in QtCreator to search a type that is found in 490 files... it takes like 10 seconds, but if you have to do that fully parsing the TU of those 490 files it would take too much (and you can't do otherwise if you don't use KeepGoing because a missing include won't return a TU as you said).

**Comment by eidheim** [comment]
There is a rather significant problem with my solution. Given the following files:
```c++
// test.hpp
#pragma once

#include <string>

class Test {
public:
  std::string testing() { return "testing"; }
};
```
Imagine that one wants to rename `Test::testing` to `Test::testing2`. One builds a PCH from the following source:
```c++
#include "test.hpp"
```
But in the following file, in order to use the same PCH, not all occurrences of Test::testing() can be renamed due to errors in the line containing `std::cout`:
```c++
// main.cpp

// #include <iostream> removed in order to use the same PCH
#include "test.hpp"

int main() {
  Test test;
  test.testing();              // test and testing can be renamed through libclang
  std::cout << test.testing(); // test and testing can NOT be renamed through libclang
  test.testing();              // test and testing can be renamed through libclang
}
```
I've tried various libclang flags trying to solve this, but no success yet.

**Comment by Smjert** [comment]
I've tested what we were talking about yesterday:
1) Took a .cpp with a type i wanted to search (KoColor in Krita source)
2) Fully parsed the TU
3) Got the USR of the type
4) Took a source file where i knew there were some references
5) Cleaned up the command line removing all "-I and -isystem" arguments
6) Parsed the TU with KeepGoing
7) Visited (recursing) all the TU children

The parsing of this last TU takes like 20ms.
It does find **part** of what i'm searching (funnily it seems also to still have some pieces from the header it includes), though there's a couple of lines like these:

```
59     KoColor currentColor;
60     KoColor buddyColor;
```

It finds line 60 but not 59.

And this is not me borking something because even with **clang-check -ast-print** (without giving the compilation database) it doesn't show it, only line 60 is shown.

But there's also another issue still... the damn #defined and #ifdefs could hide from the AST a lot of things if we don't fully parse the TUs, so you would leave a lot of pieces behind..

And apparently you kind of have the same issue i've seen.

**Comment by eidheim** [comment]
One solution that could work for up to medium sized projects is that one includes all the needed headers in the PCH. For instance, given the files:
```c++
#pragma once

#include <string>

// test.hpp

class Test {
public:
  std::string testing() { return "testing"; }
};
```
```c++
#include <iostream>
#include "test.hpp"

// main.cpp

int main() {
  Test test;
  std::cout << test.testing();
}
```
```c++
#include <vector>
#include "test.hpp"

// main2.cpp

int main() {
  std::vector<Test> test;
  test.emplace_back();
  auto testing=test.back().testing();
}
```
Then one would create a PCH from the following source:
```c++
#include <iostream>
#include <vector>
#include "test.hpp"
```
And in this source, given that one wants to rename `Test::testing()`, add the needed function bodies that include `testing` and rename these through libclang.

But again, this would potentially create a pretty large memory consuming PCH, but it would work for small and medium sized projects I think.

**Comment by Smjert** [comment]
The issue with that though are compilation flags and defines (which you can't remove in any way): to be sure that everything will work (the TU will successfully parse or you don't get a weird result) you will have to have a PCH for each set of flags/defines used in source files.

Which means that even if between A.cpp and B.cpp you actually include the same headers, so you think the same PCH can be used, if to compile A.cpp you use different flags than B.cpp, which "changes the content" of the header (-D, some -f switch), you would need a PCH for A and one for B, which means doubling the time already to create it.
Moreover you would have to rebuild both PCH from scratch at each header change.

Also i didn't get if you meant to have a single PCH for the entire project or one for each TU.. the first case has the issue that is slower to create and it might not even be correct, again for different sets of flags between source file compilations... the second one suffer from poor reusability and it would be the same thing as just parsing all the TU the first time and caching them on disk.

I'm currently trying the third "crazy way", which though i don't know how much it will cost in terms of in memory management (so it might be unfeasible this too): i'm trying to play with the frontend API of clang which lets you insert hooks when the preprocessor process the source file.
It's the original idea: build a tree of includes, store all the defines for each header which could change what the header contains (so anything that is used in #if and #ifdef) and then listen for those changes when the user writes the code or changes compilation flags.
It would still need to duplicate headers depending on flag but on a smaller scale and with less space consumption, because by creating smaller PCHs that could be used as building blocks for the final source file needed PCH you can reuse much more.

The real unknown part is how it performs between marking all the possibly dirty includes (all the parents of the changed include, plus all the one that comes after which are on the same header file level) and chaining possibly hundreds of already built PCHs vs just rebuilding from scratch.

**Comment by eidheim** [comment]
I meant 1 PCH for all the source files that maybe contain (after a grep or the like) the variable or function you a looking for. One can luckily find the flags used for each source file in `compile_commands.json`, so in cases where more than one PCH is needed, it is possible to take this into account. 

I think your "crazy way" option is interesting, and we might have to look for such options! Looking forward to see how this is going.

**Comment by Smjert** [comment]
> I meant 1 PCH for all the source files that maybe contain (after a grep or the like) the variable or function you a looking for. One can luckily find the flags used for each source file in compile_commands.json, so in cases where more than one PCH is needed, it is possible to take this into account.

Well yes the flags are in the JSON, i'm just worried about the amount of combinations.
Though even with that, it's easy to include headers that conflicts in the PCH, unless you really mean to have a preprocessor that tells you which includes are really included for that source file and set of flags. And to know that you would have to parse all the chain of includes because the #if/#ifdef that control what it has to be included might depend on something that is defined way down the chain.
I mean, sure this can work with the new small project of some guy X, but if you're writing even a "small" library that though has to deal with platform differences, so "it needs" to use ifdefs.. then i fear it's not that viable.

**Comment by eidheim** [comment]
It might be that one can do an initial parse of a source file with all the includes removed (this parse is very fast), and still be able to figure out which conditional inclusions (`#if`, `#ifdef`, ...) are valid. Might be worth investigating.

**Comment by eidheim** [comment]
From experimenting I found that include guards (not `#pragma once`) forces PCH's to be reconstructed and greatly increase the parse time of header files! I'll add removal of these include guards before passing header files to libclang in the following week hopefully. Nevertheless, this discussion most likely lead to a vast performance boost when working on complex headers:)

**Comment by Smjert** [comment]
> From experimenting I found that include guards (not #pragma once) forces PCH's to be reconstructed and greatly increase the parse time of header files![...]

What you mean exactly that it forces PCH's to be reconstructed? How are you compiling them in the first place?
Seems really weird because i would actually expect them to make the PCH compile in the first place, due to the possible multiple inclusion issue.

**Comment by eidheim** [comment]
When reusing a PCH, either internally through libclang, or an external PCH file, include guards forces a full reparse of all its header files when changes anywhere in the source is made.

**Comment by Smjert** [comment]
> When reusing a PCH, either internally through libclang, or an external PCH file, include guards forces a full reparse of all its header files when changes anywhere in the source is made.

I still don't get it or why it should happen... and why it shouldn't break when you remove the guards.
Something tells me that this is related to https://reviews.llvm.org/D15994.

**Comment by eidheim** [comment]
Yes, the patch you link to seems to be related. Currently, we now remove these include guards when one is working on a header file in juCi++. The speed up is significant. My guess though is that people will be using `#pragma once` more and more as it has been supported quite a while by both gcc (since 3.4) and clang. Though it will take a while before the C++ literature is updated, even my own lectures and presentations for NTNU are outdated...

**Comment by Smjert** [comment]
I'm slowly coming to think again that what i'm trying to do is too complicated with the tools i have.
I managed to simply have the clang preprocessor to pass on all files of a compilation database and execute some callbacks when it encounters ifs/includes, when it enters or exits from files and the like (http://clang.llvm.org/doxygen/classclang_1_1PPCallbacks.html).
In these callbacks i'm not currently doing anything and doing a single thread run on all the files takes ~300 seconds, which would mean ~50 seconds if i use my 6 cores, with 6 threads (again ~2800 files.. my target is still QtCreator which is ready in around 30 seconds), though all those seconds just to parse... and i would still have to actually store informations about defines i encounter.. build the include graph and compile/compose the PCHs.

So... i would gain some speedup if i was able with clang to avoid preprocessing headers that i've already preprocessed with the same identical set of defines... but there's no way to do that unless i reimplement some good parts of it, so that i can load from disk the internal status of the preprocessor as if it had already preprocessed those files.

Given that this means basically giving a good alternative to modules, and so partially solve the big issue in reparsing with normal compilations... it means modules aren't useful anymore and i'm a genius (sure...).

Joking aside and long story short i have no idea how to turn this into something simpler and speed up the whole thing enough.

I can already create PCH or even PTH if i want with the first source file (and all its dependency chain) i preprocess; the issue though is that for the next source file i know if a certain PCH/PTH X is good for the header Y i'm about to preprocess only after preprocessing it up to that header... and it would be already too late because i have to pass PCH/PTH to the preprocess invocation on the source file, so before starting to preprocess.

**Comment by eidheim** [comment]
First, sorry for the confusion, I meant a precompiled preamble and not PCH in my last comments. They are basically the same, but libclang works with its own precompiled preamble, and you can combine this with a PCH that is passed to libclang when parsing a source. That is, when I said reparsing the PCH, I meant reparsing the precompiled preamble of a source file. 

By the way, as I understand it, the precompiled preamble in libclang exists for faster reparsing of a source file, while PCH's exists for faster parsing of source files. 

Thank you for investigating your solution further, I still think your idea is worth investigating. I have tested using PCH's passed to libclang and it's working quite ok. One possible issue with having many PCH's, however, is that parsing of all these PCH's might take some time. 

I'll continue with my idea as well, that is working on a common precompiled preamble within libclang for the source files in question when finding function or variable usages. You would of course have the issue with conditional inclusions, but at least initially we can be working towards a 90-95% correctly working solution through simply adding the conditional inclusions when creating the precompiled preamble. 

I would not worry too much about the processing time compared to for instance Qt Creator. They have an entirely different approach, which require a significant effort to make it work for newer C++ standards. And it will most likely never reach the quality of libclang processing. 

And, of course, a libclang solution would work much faster when using C++17 modules:) Modules being basically a general approach to PCH's if I understand it correctly.
